Cropping consistency and Gravity. Allow changing aspect ratio of cropped images

mickenorlen opened the pull request:
# Update 7/7

See comment: https://github.com/AlchemyCMS/alchemy_cms/pull/1786#issuecomment-654788935

---

## What is this pull request for?
Closes #1783

- Add consistency to cropping regardless if crop mask is applied or not or using srcset.

- **Render cropping**. Allow changing aspect ratio of cropped images in render if `render_crop=true` - with added `gravity` settings (see screenshot doc for render cropping in action).

- **Gravity**. The gravity of an `essence_picture` can now be specified in content settings (**1.** for devs) image properties (**2.** for users - if `present?` in settings) or directly in render (**3.**). It can be read with proper priority, validation, fallbacks to defaults etc through `essence.gravity` method. Eg. if you want to use it for aligning background images or something other than its natural use in render_cropping.

- **Screenshot doc** with explanations and tests of render cropping in action 
https://docs.google.com/document/d/1rRCEBYSnKczOS3HYNHFdy2s-EzY17B0G3ljp1vLMLis/edit?usp=sharing


- **TODO** 
- [x] Add `render_gravity` column to cms. **I added the migration 20200701100718**
- [x] I have added tests to cover this change
- [X] I have followed [Pull Request guidelines](https://github.com/AlchemyCMS/alchemy_cms/blob/master/CONTRIBUTING.md)
- [X] I have added a detailed description into each commit message (Not so much but lots of comments in code)

```RUBY
class AddGravityToEssencePicture < ActiveRecord::Migration[6.0]
  def change
    add_column :alchemy_essence_pictures, :render_gravity, :string
  end
end

```


### 1. Gravity in content settings

 `elements.yml`

```YML
# By default there's no visible change of this PR
# If no gravity is specified it won't be visible to users
# If render_crop != true gravity won't have any effect anyway
# It is only applied when the image aspect ratio changes
- name: no_gravity_visible
  type: EssencePicture

# With default gravity visible in page properties
- name: default_gravity_image
  type: EssencePicture
  settings:
    gravity: true

# With gravity visible in page properties and overwritten layout_defaults
- name: custom_gravity_image
  type: EssencePicture
  settings:
    gravity:
      size: shrink,  # default = grow
      x: left,   # default = center
      y: top   # default = center
```

### 2. Gravity in page properties
I serialized the `render_gravity` (string in db) as JSON so its working as a hash. Then I had to split the params in three and combine them in controller:

`essence_pictures_controller.rb`

```RUBY
def combine_gravities(essence_picture_params)
  gravity = @essence_picture.render_gravity || {}
  gravity = gravity.symbolize_keys.merge({
    size: essence_picture_params.delete(:render_gravity_size),
    x: essence_picture_params.delete(:render_gravity_x),
    y: essence_picture_params.delete(:render_gravity_y)
  })

  essence_picture_params[:render_gravity] = gravity.reject{ |_k,v| v.blank? }.presence || nil
  essence_picture_params
end
```

![image](https://user-images.githubusercontent.com/9784434/80286996-1c865600-872f-11ea-9d52-981a9b714737.png)

### 3. Gravity in render
```RUBY
<%= el.render(:blog_post_image, size: '600x400', render_crop: true, gravity: { size: 'shrink' }) %>
```
---

mickenorlen commented:
I think i will have to distinguish between `crop` and `render_crop` so it's still possible to resize normally (without cropping) a cropped image. Ill update the code with this later..

tvdeyen commented:
Thanks. And sorry for letting you wait so long. 

Unfortunately I am not sure I fully understand what this rather complex feature is meant for. I read the Google doc. Thanks for that. 

Am I assuming correctly that you want to use these values to align the image as `background-image` and set the `background-position` and `background-size` values in CSS accordingly?

I think this is overall a good idea, but the code is very(!) complex. Without tests and more example code (especially on the frontend) I am not able to review it. 

Could you provide more examples and tests?

Thanks for your contribution

mickenorlen commented:
Hey, I'm currently on the road so can't give you a more detailed reply until in a week or so.

> I think this is overall a good idea, but the code is very(!) complex. Without tests and more example code (especially on the frontend) I am not able to review it.

I understand this PR needs more work to be properly reviewed. But I'm glad to have your first opinion so we can start to discuss if this patch has the potential to be merged and if so, how to move forward with it.

> Am I assuming correctly that you want to use these values to align the image as background-image and set the background-position and background-size values in CSS accordingly?

The main use cases for this feature. 

- **Responsive images with dynamic aspect ratio**. I'm using the picture tag with srcset to load different image versions depending on screen size. https://www.w3schools.com/tags/tag_picture.asp. This is particularly useful for wide images like banners or the like.
- **Images used in multiple contexts**. Consider a blog post element that has a "main image" content. 
-- First of all I want to force the user to crop the image to a standardized size for display at the introduction of the blog post page to maintain a consistency between posts. For this I'm using the current crop feature with a custom min-size validation.
-- Then I want to be able to use that same "main image" in the blog post list/grid views. For my design I want to present it in a different aspect ratio (less wide). 
-- Then I also want it to pop up in search autocompletion... 
-- It also just so happens that this "main image" is presented as quite large and wide... so I would like it to be less wide in smaller screens.

So basically this feature helps me force standardized image sizes while avoiding having users manually crop 4 different versions of the same image.

mickenorlen commented:
# General feature logic

Picture aspect ratio is determined in the following order:

1. `size` (or srcset-size) with `render_crop = true` (applied from render)
2. `crop_from`/`crop_size` (applied by user in cropping tool)
3. `render_size` (selected by user from `sizes` options)
4. `size` (or srcset-size) without `render_crop`

Now you can proceed to the test guide! 

(If you want more examples about how `render_crop` and `gravity` works, see doc: https://docs.google.com/document/d/1rRCEBYSnKczOS3HYNHFdy2s-EzY17B0G3ljp1vLMLis/edit#)


# Test guide

## 1. Download 720x720 color square image: 
https://cdn.pixabay.com/photo/2013/10/21/04/51/color-198892_960_720.jpg
## 2. Create pages / elements
`page_layouts.yml`
```YML
- name: render_crop_page
  elements:
    - user_crop_element
    - render_size_element
```

`elements.yml`
```YML
- name: user_crop_element
  contents:
  - name: user_crop_picture
    type: EssencePicture
    hint: 'Using crop_from/crop_size for user cropping'
    settings:
      gravity: true
      crop: true
      size: 400x200

- name: render_size_element
  contents:
  - name: render_size_picture
    type: EssencePicture
    hint: 'Using render_size for user size selection'
    settings:
      # Setting a new default gravity here
      gravity: {
        x: 'left'
      }
      crop: true
      size: 400x200
      sizes:
        - ["Small, 200x200", "400x400"]
        - ["Tall, 200x400", 200x400]

- name: srcset_element
  contents:
  - name: srcset_picture
    type: EssencePicture
    hint: 'Using srcset'
    settings:
      gravity: true
      crop: true
      size: "600x300"
      srcset: ['200x100', '400x200', '600x300']
      sizes: ['(min-width: 600px) 600px', '100vw']
```

`elements/_user_crop_element.html.erb`

```ERB
<%- cache(user_crop_element) do -%>
  <h1>User crop element</h1>
  <%= element_view_for(user_crop_element) do |el| -%>
    <%= el.render :user_crop_picture %>
  <%- end -%>
<%- end -%>
```

`elements/_render_size_element.html.erb`
```ERB
<%- cache(render_size_element) do -%>
  <h1>Render size element</h1>
  <%= element_view_for(render_size_element) do |el| -%>
    <%= el.render :render_size_picture %>
  <%- end -%>
<%- end -%>
```

`elements/_srcset_element.html.erb`
```ERB
<%- cache(srcset_element) do -%>
  <h1>Srcset element</h1>
  <%= element_view_for(srcset_element) do |el| -%>
    <%= el.render :srcset_picture, {}, style: { width: "100vw", max_width: '600px'} %>
  <%- end -%>
<%- end -%>
```

## 3. Go in to admin and create the page with its elements

![image](https://user-images.githubusercontent.com/9784434/86760608-a151f200-c045-11ea-9f44-b3191c93e536.png)

---

## 4. Test User crop element

`elements/_user_crop_element.html.erb`
- Change size in render (should change crop aspect ratio):
`<%= el.render :user_crop_picture, { size: "200x400"} %>`
- Open crop tool and apply default crop mask. Should now override aspect ratio of requested size from render as that size is only used to resize a user cropped area.
- Apply render_crop (aspect ratio should now fit requested size again - growing the crop area by default).
` <%= el.render :user_crop_picture, { size: "200x400", render_crop: true } %>` 

## 5. Test Render size element
`elements/_render_size_element.html.erb`
- Change size in render (should change crop aspect ratio)
`<%= el.render :render_size_picture, { size: '200x200' } %>`
- Change `render_size` to 200x400 in picture properties and update view to see change. (should override size unless set to layout default)
- Open the cropping tool and move crop area to the right and apply. Should apply properly.
- Apply render_crop
`<%= el.render :render_size_picture, { size: '200x200', render_crop: true } %>`
![image](https://user-images.githubusercontent.com/9784434/86767332-1aecde80-c04c-11ea-9e15-c5d52c663865.png)
Now image didnt fully grow back to normal because we have `gravity: {x: left}` default in settings. Change gravity x to `right` in picture properties or render and you should have a full square again. As the right side keeps its position/alignment, the left side can grow.

## 6. Test Srcset element
`elements/_srcset_element.html.erb`
- Test that the source image is updated according to specified srcset in mobile screen size.
- Change size in render. Should not update aspect ratio as size is only used for crop area sizing in this case.
`<%= el.render :srcset_picture, { size: "300x600" }, style: { width: "100vw", max_width: '600px'} %>`
- Change srcset. Should update sizes/aspect ratios:
`<%= el.render :srcset_picture, { srcset: ['100x200', '200x400', '300x600'] }, style: { width: "100vw", max_width: '600px'} %>`
- Now use cropping tool to apply a user defined crop area. Should apply, now the srcset sizes only resize instead of fit aspect ratio.
- Set render_crop: true. Should force srcset sizes to determine aspect ratio.
`<%= el.render :srcset_picture, { srcset: ['100x200', '200x400', '300x600'], render_crop: true }, style: { width: "100vw", max_width: '600px'} %>`

tvdeyen commented:
I am so sorry that I still haven't time to look into this. I still think it is good, but need time to test and review this. Do you think there might be some somit that could be extracted into separate PRs that are smaller and easier to review? That might help to get this in easier. 

Again, thank you for this great work.

mickenorlen commented:
> I am so sorry that I still haven't time to look into this. I still think it is good, but need time to test and review this. Do you think there might be some somit that could be extracted into separate PRs that are smaller and easier to review? That might help to get this in easier.
> 
> Again, thank you for this great work.

Alright, I split out https://github.com/AlchemyCMS/alchemy_cms/pull/1935 and https://github.com/AlchemyCMS/alchemy_cms/pull/1934

I saw your implementation of `calculations.rb` and moved some logic there. The test notes above still apply.

tvdeyen commented:
This probably needs a rebase now that we merged some of the extracted bug fixes.

mickenorlen commented:
Rdy for review

tvdeyen commented:
@mickenorlen we switched the default branch to `main`. I already updated this PR, but you want to pull the `main` branch and rebase this branch against this as well, so you get future updates. sorry for any inconvenience and thanks for the contribution.

mickenorlen commented:
@tvdeyen Force pushed a fresh single squash merge commit into main as rebase was a mess.

mickenorlen commented:
I came up with a few small changes that might make this PR a bit easier for you to review. For example breaking out a new `render_crop.rb` module.. I'll get back to this in a couple of days.

mickenorlen commented:
@tvdeyen, After refactoring most logic to `alchemy/picture/render_crop.rb` (and adding tons of general comments about the feature there) everything came out much cleaner. I also found some general code quality improvements. Should be easier to review now, when you have time. Thanks

tvdeyen commented:
This is still relevant and and useful feature that we would love to have, but this PR needs to be reworked from scratch since we made a lot of refactorings lately.

tvdeyen commented:
Maybe this can be split up into smaller PRs? One that introduces the gravity feature and one that handles aspect ratio changes.

mickenorlen commented:
I've recently cleaned up the code behind this feature a bit as i used it in my own project.
I've actually renamed "gravity" to "focus" in my new version as i felt it was more intuitive for a user eg: 

"The focus in this picture is: x: left/center/right, y: top/center/bot, zoom: in/closest/out" .   

Then the developer can make render choices based on this - cropping can be one of them, background-align another etc.


In short i've made a `Render::Crop` class that takes a wanted `size` (or `aspect_ratio`) and changes the cropping area to fit.


```RUBY
class Render::Crop
  def initialize(size:, original_size:, crop_from:, crop_size:, focus:, aspect_ratio: nil)
    ...
    @crop_from = to_crop_from(crop_from) ||  { x: 0, y: 0 }
    @crop_size = to_crop_size(crop_size) || @original_size
  end

  def adjust_crop_area_to_aspect_ratio
    new_crop_size = case @focus[:zoom]
    when :out, :closest
      grow_crop_area
    when :in
      shrink_crop_area
    end

    crop_from = crop_from_after_crop_area_resize(new_crop_size)

    [crop_from, new_crop_size]
  end    
  
  ...
  
  private
  
  # The wanted aspect ratio is taken either from size or a directly from aspect ratio
  def wanted_aspect_ratio
    @wanted_aspect_ratio ||= (@aspect_ratio.presence || aspect_ratio(@size))
  end
end
```

One thing that was useful for me to implement related to this feature was to make picture `size` and `aspect_ratio` methods that always returns something, falling back correctly through render->render_size->settings->original_image. Then you just pass them in to the render_cropper which returns a new crop area. I even used the render_cropper to get default_crop_areas for admin thumbnails etc


Can you give me your thoughts on all this and a quick explanation of how your recent refactorings influence this feature?

tvdeyen commented:
I like the idea of a render/crop focus class. Could you elaborate what the values mean? Why is zoom in and zoom out a focus related value? I would rather think of a coordinate instead of a verb/command, but I probably miss some context.

mickenorlen commented:
@tvdeyen 
> I like the idea of a render/crop focus class. Could you elaborate what the values mean? Why is zoom in and zoom out a focus related value? I would rather think of a coordinate instead of a verb/command, but I probably miss some context.

The focus zoom property is perhaps a bit unintuitive and I'm open for suggestions for another wording. 

In general `focus[:zoom]` is specifying the strategy to achieve the wanted aspect ratio: 
- `out`: By zooming out, effectively "growing" the cropped image horizontally or vertically.
- `in`: by zooming in, effectively "shrinking" the image mask in the opposite direction compared to growing
- `closest`: Growing one side and shrinking the other 50/50.

With `focus[:x] = left` the left focus/coordinate of the cropping mask is maintained and the new mask would only grow/shrink on the right if horizontal adjustments are required. Eg if shrinking a wider image, or growing a taller one.

With `focus[:y] = top` the top focus/coordinate of the cropping mask is maintained and the new mask would only grow/shrink on the bottom if vertical adjustments are required. Eg: if shrinking a taller image, or growing a wider one.

I've updated the comments in the `Render::Crop` file so you can see the whole class here: https://pastebin.com/78ZRKutk.

In the `self.crop` method you can see a bit how I prepare the parameters to call the main method `adjust_crop_area_to_aspect_ratio`. The `wanted_aspect_ratio` is either derived directly from the `aspect_ratio` parameter or from the `size` parameter eg: `"400x200"`.
